### PR TITLE
TDL-19860: Need to update Primary keys for Company_Attributes & Contacts streams

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-19860: Need to update Primary keys for Company_Attributes & Contacts streams
- Updated the Primary Key for `company_attributes` and used the hash of `['id', 'name', 'description']` as the Primary Key.

**NOTE:** The `contact_attributes` is also using the same endpoint as the `company_attributes` thus, updated the Primary Key for contact_attributes as well.

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_intercom/streams.py [hunk 1]
```diff
@@ -63,6 +65,27 @@ def get_parent_data(self, bookmark_datetime: datetime = None) -> list:
...
+        """
+        # There are 2 types for data_attributes in Intercom
+        # -> Default: As discussed with support, there is 'id' for custom data_attributes and ther will be unique
+        # -> Custom: Used 'name' and 'description' for identifying the data uniquely
+        fields_to_hash = ['id', 'name', 'description']
+        hash_string = ''
+
+        for key in original_record:
+            if key in fields_to_hash:
+                hash_string += str(original_record.get(key, ''))
```
Comment (savan-chovatiya) [inline tap_intercom/streams.py:80]:
```suggestion
        for key in fields_to_hash:
                hash_string += str(original_record.get(key, ''))
```

File: tap_intercom/streams.py [hunk 2]
```diff
@@ -160,32 +183,53 @@ def sync(self,
...
+            activate_version_message = singer.ActivateVersionMessage(
+                stream=self.tap_stream_id,
+                version=activate_version)
+            singer.write_message(activate_version_message)
 
         with metrics.record_counter(self.tap_stream_id) as counter:
             for record in self.get_records():
+
+                if self.tap_stream_id in ['company_attributes', 'contact_attributes']:
+                    record = self.generate_record_hash(record)
```
Comment (savan-chovatiya) [inline tap_intercom/streams.py:211]: Add code comment

File: tap_intercom/streams.py [hunk 3]
```diff
@@ -496,11 +546,17 @@ class ContactAttributes(FullTableStream):
...
+    key_properties = ['_sdc_record_hash']
     path = 'data_attributes'
     params = {'model': 'contact'}
     data_key = 'data'
 
+    def sync(self, state: dict, stream_schema: dict, stream_metadata: dict, config: dict, transformer: Transformer, sync_with_version: bool = False) -> dict:
+        # Sync with activate version
+        # As we are preparing the hash of ['id', 'name', 'description'] and using it as the Primary Key and there are chances
+        # of field value being updated, thus, on the target side, there will be the redundant entry of the same record.
+        return super().sync(state, stream_schema, stream_metadata, config, transformer, sync_with_version = True)
```
Comment (savan-chovatiya) [inline tap_intercom/streams.py:558]: Remove this function and define one class attribute in the base class as `sync_with_version = False`, and override it with `sync_with_version  = True` for 2 streams.

File: tests/unittests/test_sdc_record_hash.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    return args[0]
+
+class TestSDCRecordHash(unittest.TestCase):
+    @parameterized.expand([
+        ['default_attr', {'name': 'test 1', 'description': 'test 1 description', 'label': 'test1'}],
+        ['default_attr_without_description', {'name': 'test 1', 'label': 'test1'}],
+        ['custom_attr', {'id': 11, 'name': 'test 1', 'description': 'test 1 description', 'label': 'test1'}],
+        ['custom_attr_without_description', {'id': 11, 'name': 'test 1', 'label': 'test1'}],
+    ])
+    def test_generate_hash(self, name, test_data):
```
Comment (savan-chovatiya) [inline tests/unittests/test_sdc_record_hash.py:17]: Add doc string for all unit test cases.

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tap_intercom/streams.py [hunk 5]
```diff
@@ -63,6 +65,27 @@ def get_parent_data(self, bookmark_datetime: datetime = None) -> list:
         parent = self.parent(self.client)
         return parent.get_records(bookmark_datetime, is_parent=True)
 
+    def generate_record_hash(self, original_record):
+        """
+            Function to generate the hash of name, full_name and label to use it as a Primary Key
+        """
+        # There are 2 types for data_attributes in Intercom
+        # -> Default: As discussed with support, there is 'id' for custom data_attributes and ther will be unique
```
Comment (dbshah1212) [inline tap_intercom/streams.py:73]:
```suggestion
        # -> Default: As discussed with support, there is an 'id' for custom data_attributes and that will be unique
```

File: tap_intercom/streams.py [hunk 6]
```diff
@@ -160,32 +183,53 @@ def sync(self,
...
                 transform_times(record, schema_datetimes)
 
                 transformed_record = transform(record,
                                                 stream_schema,
                                                 integer_datetime_fmt=UNIX_MILLISECONDS_INTEGER_DATETIME_PARSING,
                                                 metadata=stream_metadata)
                 # Write records with time_extracted field
-                singer.write_record(self.tap_stream_id, transformed_record, time_extracted=singer.utils.now())
+                if sync_with_version:
+                    # Using "write_message" if the version is found as, "write_record" params does not contain "version"
```
Comment (dbshah1212) [inline tap_intercom/streams.py:221]:
```suggestion
                    # Using "write_message" if the version is found. As "write_record" params do not contain "version"
```

Comment (harshpatel4crest) [reply tap_intercom/streams.py]: Updated the comment.

File: tap_intercom/streams.py (hunk 5 above)
Comment (harshpatel4crest) [reply tap_intercom/streams.py]: Updated the comment.

File: tests/unittests/test_sdc_record_hash.py (hunk 4 above)
Comment (harshpatel4crest) [reply tests/unittests/test_sdc_record_hash.py:17]: Added function comment for the tests.

File: tap_intercom/streams.py (hunk 3 above)
Comment (harshpatel4crest) [reply tap_intercom/streams.py]: Updated the code with the suggested approach.

File: tap_intercom/streams.py (hunk 2 above)
Comment (harshpatel4crest) [reply tap_intercom/streams.py:212]: Added comment.

File: tap_intercom/streams.py (hunk 1 above)
Comment (harshpatel4crest) [reply tap_intercom/streams.py]: Updated the code.

**Review by savan-chovatiya — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

File: tests/test_intercom_automated_fields.py [hunk 7]
```diff
@@ -54,7 +54,8 @@ def test_run(self):
```
Comment (RushiT0122) [inline tests/test_intercom_automated_fields.py:54]: Standard assertions missing for automatic fields.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Standard assertions missing from automatic fields integration test.

**Comment by harshpatel4crest** [comment]
> Standard assertions missing from automatic fields integration test.

@RushiT0122 They are added as part for PR #57

**Review by umangagrawal-crest — CHANGES_REQUESTED** [review]
Please update README file

**Review by savan-chovatiya — APPROVED** [review]

**Comment by harshpatel4crest** [comment]
> Please update README file

Updated README file.

File: tests/test_intercom_automated_fields.py [hunk 8]
```diff
@@ -54,7 +54,8 @@ def test_run(self):
 
                 # collect actual values
                 data = synced_records.get(stream, {})
-                record_messages_keys = [set(row.get('data').keys()) for row in data.get('messages', {})]
+                record_messages_keys = [set(row.get('data').keys()) for row in data.get('messages', {})
```
Comment (RushiT0122) [inline tests/test_intercom_automated_fields.py:57]: - Name of test file should be `test_intercom_automatic_fields.py`
- Assertion missing `Verify that all replicated records have unique primary key values.`
- Doc string should be fixed

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Requested changes in integration test

File: tests/test_intercom_automated_fields.py (hunk 8 above)
Comment (harshpatel4crest) [reply tests/test_intercom_automated_fields.py:57]: They are covered as part of Tap-Tester tests PR #57

**Review by RushiT0122 — APPROVED** [review]

**Review by luandy64 — APPROVED** [review]

**Pull request merged**